moto/sns/models.py

```diff
@@ -832,6 +832,72 @@ def _validate_filter_policy(self, value):
                             )
                         continue
                     elif keyword == "numeric":
+                        # TODO: All of the exceptions listed below contain column pointing where the error is (in AWS response)
+                        # Example: 'Value of < must be numeric\n at [Source: (String)"{"price":[{"numeric":["<","100"]}]}"; line: 1, column: 28]'
+                        # While it probably can be implemented, it doesn't feel as important as the general parameter checking
+
+                        from xml.sax.saxutils import escape
```

Comment:
This function is used to escape XML (`<`, `>`, etc. characters) in the exception thrown, but I feel like it's not the best place to do it. Is there a better way to put it so it works globally?

Without this in place, boto client errors out on parsing the response, due to the forbidden characters.

Reply:
The `RESTError`-class is the ancestor for all exceptions, so that's probably the best choice: 
https://github.com/getmoto/moto/blob/462ff57900abc0a8133a26487052273f9fc4579f/moto/core/exceptions.py#L58

Reply:
I am not sure if we understood each other. The issue here is IMHO not that the exception is raised on moto's end (so we can catch the generic `RestError`), but rather that moto happily returns a message that can't be parsed by the boto client.

So it seems like there's nothing on the moto's end making sure that content is escaped before being sent out and it can't really be catched with an exception. I am not completely sure how the moto/boto interaction works though, so I might be wrong.

If it's OK for you then keeping the sanitizing on the local level should be OK, most of the responses won't ever contain problematic characters.

Here's the trace if the `escape`s are removed:
```

        with pytest.raises(ClientError) as err:
>           conn.subscribe(
                TopicArn=topic_arn,
                Protocol="http",
                Endpoint="http://example.com/",
                Attributes={
                    "FilterPolicy": json.dumps({"price": [{"numeric": ["<", "100"]}]})
                },
            )

tests/test_sns/test_subscriptions_boto3.py:484: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
.venv/lib/python3.10/site-packages/botocore/client.py:530: in _api_call
    return self._make_api_call(operation_name, kwargs)
.venv/lib/python3.10/site-packages/botocore/client.py:943: in _make_api_call
    http, parsed_response = self._make_request(
.venv/lib/python3.10/site-packages/botocore/client.py:966: in _make_request
    return self._endpoint.make_request(operation_model, request_dict)
.venv/lib/python3.10/site-packages/botocore/endpoint.py:119: in make_request
    return self._send_request(request_dict, operation_model)
.venv/lib/python3.10/site-packages/botocore/endpoint.py:199: in _send_request
    success_response, exception = self._get_response(
.venv/lib/python3.10/site-packages/botocore/endpoint.py:241: in _get_response
    success_response, exception = self._do_get_response(
.venv/lib/python3.10/site-packages/botocore/endpoint.py:308: in _do_get_response
    parsed_response = parser.parse(
.venv/lib/python3.10/site-packages/botocore/parsers.py:250: in parse
    parsed = self._do_error_parse(response, shape)
.venv/lib/python3.10/site-packages/botocore/parsers.py:557: in _do_error_parse
    root = self._parse_xml_string_to_dom(xml_contents)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <botocore.parsers.QueryParser object at 0x7f0c29d9d9f0>
xml_string = b'<?xml version="1.0" encoding="UTF-8"?>\n<ErrorResponse>\n    <Error>\n        <Code>InvalidParameter</Code>\n       ...Message>\n        \n        <RequestId>7a62c49f-347e-4fc4-9331-6e8eEXAMPLE</RequestId>\n    </Error>\n</ErrorResponse>'

    def _parse_xml_string_to_dom(self, xml_string):
        try:
            parser = ETree.XMLParser(
                target=ETree.TreeBuilder(), encoding=self.DEFAULT_ENCODING
            )
            parser.feed(xml_string)
            root = parser.close()
        except XMLParseError as e:
>           raise ResponseParserError(
                "Unable to parse response (%s), "
                "invalid XML received. Further retries may succeed:\n%s"
                % (e, xml_string)
            )
E           botocore.parsers.ResponseParserError: Unable to parse response (not well-formed (invalid token): line 5, column 79), invalid XML received. Further retries may succeed:
E           b'<?xml version="1.0" encoding="UTF-8"?>\n<ErrorResponse>\n    <Error>\n        <Code>InvalidParameter</Code>\n        <Message>Invalid parameter: Attributes Reason: FilterPolicy: Value of < must be numeric\n at ...</Message>\n        \n        <RequestId>7a62c49f-347e-4fc4-9331-6e8eEXAMPLE</RequestId>\n    </Error>\n</ErrorResponse>'

.venv/lib/python3.10/site-packages/botocore/parsers.py:506: ResponseParserError
```

Reply:
I was talking about a Moto-class called `RESTError`, that contains all the logic for exceptions to be properly formatted before being send to boto

Reply:
Ok, I got it now, thanks. I'll move the logic there

Reply:
Ended up adapting the logic from https://github.com/getmoto/moto/commit/579b6e264ff297d1787706848f005b5d707bb794 for all core error templates to avoid double-escaping (as would be the case with SES after original changes).
CDATA seems a lot cleaner anyway